libglusterfs/src/gf-io.c

```diff
@@ -0,0 +1,1626 @@
+    "instability or hangs, specially during process termination."
+
+#define LG_MSG_IO_THREAD_FAILED_LVL(_res) GF_LOG_ERROR
+#define LG_MSG_IO_THREAD_FAILED_FMT                                            \
+    "Unable to create a worker thread. System may be extremely low on "        \
+    "resources."
+
+#define LG_MSG_IO_WORKER_STARTED_LVL(_res) GF_LOG_DEBUG
+#define LG_MSG_IO_WORKER_STARTED_FMT "Worker %p started."
+
+#define LG_MSG_IO_WORKER_STOPPED_LVL(_res)                                     \
+    ((_res) == 0 ? GF_LOG_DEBUG : GF_LOG_WARNING)
+#define LG_MSG_IO_WORKER_STOPPED_FMT "Worker %p stopped."
+
+#define LG_MSG_IO_WORKER_FAILED_LVL(_res) GF_LOG_CRITICAL
+#define LG_MSG_IO_WORKER_FAILED_FMT                                            \
+    "Worker %p failed after multiple unrecoverable errors."
+
+#define LG_MSG_IO_WORKER_RESULT_LVL(_res)                                      \
+    ((_res) == 0 ? GF_LOG_DEBUG : GF_LOG_WARNING)
+#define LG_MSG_IO_WORKER_RESULT_FMT "Thread of worker %p terminated."
+
+#define LG_MSG_IO_STOP_LOST_LVL(_res) GF_LOG_CRITICAL
+#define LG_MSG_IO_STOP_LOST_FMT                                                \
+    "Previous processing errors made impossible to correctly handle a stop "   \
+    "request. Process may not be able to terminate."
+
+#define LG_MSG_IO_FALLBACK_LVL(_res) GF_LOG_INFO
+#define LG_MSG_IO_FALLBACK_FMT "Falling back to traditional I/O mode."
+
+#define gf_io_log(_res, _msg, _args...)                                        \
+    gf_msg("io", _msg##_LVL(_res), -(_res), _msg, _msg##_FMT, ##_args)
+
+#define gf_io_debug(_res, _fmt, _args...)                                      \
+    gf_msg_debug("io", -(_res), _fmt, ##_args)
+
+#ifdef HAVE_IO_URING
+
+/* Special gf_io_list_item_t pointer to mark availability of requests. */
+#define GF_IO_REQUEST_READY ((gf_io_list_item_t *)(intptr_t)-1L)
+
+#define GF_IO_BITNAME(_prefix, _name)                                          \
+    {                                                                          \
+        _prefix##_##_name, #_name                                              \
+    }
+
+typedef struct _gf_io_bitname {
+    uint64_t bit;
+    const char *name;
+} gf_io_bitname_t;
+
+static __thread gf_io_worker_t gf_io_worker = {};
+
+#endif /* HAVE_IO_URING */
+
+gf_io_t gf_io = {};
+
+/* Get a monotonic time. */
+static int32_t
+gf_io_sys_clock_gettime(struct timespec *now)
+{
+    int32_t res;
+
+    res = clock_gettime(CLOCK_MONOTONIC, now);
+    if (caa_unlikely(res < 0)) {
+        res = -errno;
+        gf_io_log(res, LG_MSG_IO_SYSCALL, "clock_gettime");
+    }
+
+    return res;
+}
+
+static uint64_t
+gf_io_time_ns(struct timespec *time)
+{
+    uint64_t ns;
+
+    ns = (uint64_t)time->tv_sec * 1000000000ULL;
+    ns += (uint64_t)time->tv_nsec;
+
+    return ns;
+}
+
+/* Destroy a mutex. */
+static int32_t
+gf_io_sys_mutex_destroy(pthread_mutex_t *mutex)
```

Comment:
nit: all the functions below, such as this one, is not specific to this framework? Can it be shared across the Gluster code so others can benefit from it (apart from the specific log, perhaps?)

Reply:
Yes. This should be moved to a common place to be available to all modules. It's simply a thin wrapper that logs an error message if the call fails. The only issue to consider here is that in some cases an error is expected, so it shouldn't be logged. To have this as a common implementation, there should be some way to handle these cases without requiring using other approaches for particular cases. I would like to think about how to do this before moving it. I would also like to move it as an independent patch to add more changes related to this kind of functions.

For this patch I would prefer to keep it here.